File: experiments/adrianco/experiment-75-opus55-alllang/brazil/runs/effort=low_language=go_model=claude-opus-5-5_prompt=neutral/rep1/soccer_test.go
package main

import (
	"bytes"
	"encoding/json"
	"strings"
	"sync"
	"testing"
	"time"
)

var (
	dbOnce sync.Once
	testDB *DB
	dbErr  error
)

// givenDataLoaded is the shared "Given the match data is loaded" step.
func givenDataLoaded(t *testing.T) *Server {
	t.Helper()
	dbOnce.Do(func() { testDB, dbErr = LoadDB("data/kaggle") })
	if dbErr != nil {
		t.Fatalf("load: %v", dbErr)
	}
	return NewServer(testDB)
}

func call(t *testing.T, s *Server, tool string, args Args) string {
	t.Helper()
	start := time.Now()
	out, err := s.CallTool(tool, args)
	if err != nil {
		t.Fatalf("%s(%v): %v", tool, args, err)
	}
	if d := time.Since(start); d > 2*time.Second {
		t.Errorf("%s took %v", tool, d)
	}
	return out
}

func mustContain(t *testing.T, out string, subs ...string) {
	t.Helper()
	for _, s := range subs {
		if !strings.Contains(out, s) {
			t.Errorf("output missing %q:\n%s", s, out)
		}
	}
}

func TestNormalizeTeam(t *testing.T) {
	cases := map[string]string{
		"Palmeiras-SP": "palmeiras", "Palmeiras": "palmeiras", "São Paulo": "sao paulo", "Sao Paulo-SP": "sao paulo",
		"Grêmio": "gremio", "Atlético-MG": "atletico mineiro", "Atletico Mineiro": "atletico mineiro",
		"Atlético - PR": "athletico paranaense", "Athletico-PR": "athletico paranaense", "Athletico Paranaense": "athletico paranaense",
		"Vasco da Gama-RJ": "vasco", "Vasco Da Gama RJ": "vasco", "Botafogo RJ": "botafogo", "Botafogo SP": "botafogo sp",
		"Sport Club Corinthians Paulista": "corinthians", "EC Bahia": "bahia", "Fortaleza FC": "fortaleza",
		"Flamengo - RJ": "flamengo", "Nacional (URU)": "nacional", "Red Bull Bragantino-SP": "bragantino",
	}
	for in, want := range cases {
		if got := NormalizeTeam(in); got != want {
			t.Errorf("NormalizeTeam(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestParseDate(t *testing.T) {
	for in, want := range map[string]string{"2023-09-24": "2023-09-24", "29/03/2003": "2003-03-29", "2012-05-19 18:30:00": "2012-05-19"} {
		d, _, err := ParseDate(in)
		if err != nil || d.Format("2006-01-02") != want {
			t.Errorf("ParseDate(%q) = %v, %v", in, d, err)
		}
	}
}

// Scenario: All 6 CSV files are loadable and queryable
func TestAllFilesLoaded(t *testing.T) {
	s := givenDataLoaded(t)
	want := map[string]int{"Brasileirao_Matches.csv": 4180, "Brazilian_Cup_Matches.csv": 1337, "Libertadores_Matches.csv": 1255,
		"BR-Football-Dataset.csv": 10296, "novo_campeonato_brasileiro.csv": 6886, "fifa_data.csv": 18207}
	for f, n := range want {
		if testDB.Files[f] != n {
			t.Errorf("%s: loaded %d rows, want %d", f, testDB.Files[f], n)
		}
	}
	srcs := map[string]bool{}
	for _, m := range testDB.Matches {
		srcs[m.Source] = true
	}
	if len(srcs) != 5 {
		t.Errorf("matches from %d sources, want 5", len(srcs))
	}
	mustContain(t, call(t, s, "dataset_info", Args{}), "fifa_data.csv: 18207 rows")
}

// Scenario: Find matches between two teams
func TestFindMatchesBetweenTeams(t *testing.T) {
	s := givenDataLoaded(t)
	ms := testDB.Find(MatchFilter{Team: "Flamengo", Opponent: "Fluminense"})
	if len(ms) < 20 {
		t.Fatalf("expected many Fla-Flu matches, got %d", len(ms))
	}
	for _, m := range ms {
		if m.Date.IsZero() || m.Competition == "" {
			t.Errorf("match missing date/competition: %+v", m)
		}
		pair := map[string]bool{m.HomeKey: true, m.AwayKey: true}
		if !pair["flamengo"] || !pair["fluminense"] {
			t.Errorf("wrong teams: %s vs %s", m.Home, m.Away)
		}
	}
	out := call(t, s, "head_to_head", Args{"team_a": "Flamengo-RJ", "team_b": "Fluminense"})
	mustContain(t, out, "Fla-Flu", "Head-to-head in dataset", "wins", "draws")
}

// Scenario: Get team statistics
func TestTeamStatistics(t *testing.T) {
	s := givenDataLoaded(t)
	out := call(t, s, "team_stats", Args{"team": "Palmeiras", "season": float64(2023)})
	mustContain(t, out, "Wins:", "Draws:", "Losses:", "Goals For:", "Goals Against:", "Win rate:")
	r, _ := testDB.TeamRecord(MatchFilter{Team: "Corinthians", Season: 2022, Competition: "Brasileirão", Venue: "home"})
	if r.Played != 19 || r.W+r.D+r.L != 19 {
		t.Errorf("Corinthians 2022 home record: %+v, want 19 matches", r)
	}
}

// Scenario: Standings calculated from matches; "Who won the 2019 Brasileirão?"
func TestStandings2019(t *testing.T) {
	s := givenDataLoaded(t)
	table := testDB.Standings(2019, "")
	if len(table) != 20 {
		t.Fatalf("2019 table has %d teams, want 20", len(table))
	}
	top := table[0]
	if top.Team != "flamengo" || top.Points() != 90 || top.W != 28 || top.D != 6 || top.L != 4 {
		t.Errorf("2019 champion = %+v, want Flamengo 90pts 28-6-4", top)
	}
	for _, r := range table {
		if r.Played != 38 {
			t.Errorf("%s played %d, want 38 (dedup across sources)", r.Team, r.Played)
		}
	}
	out := call(t, s, "standings", Args{"season": float64(2019)})
	mustContain(t, out, "1. Flamengo - 90 pts (28W, 6D, 4L)", "Champion", "Relegated")
}

// Scenario: Copa do Brasil finals and Libertadores stages
func TestCupFinals(t *testing.T) {
	s := givenDataLoaded(t)
	ms := testDB.Find(MatchFilter{Competition: "Copa do Brasil", Stage: "final"})
	if len(ms) == 0 {
		t.Fatal("no Copa do Brasil finals")
	}
	for _, m := range ms {
		if m.Competition != CompCopaBrasil {
			t.Errorf("unexpected competition %s", m.Competition)
		}
	}
	out := call(t, s, "search_matches", Args{"competition": "Libertadores", "season": float64(2018), "stage": "final"})
	mustContain(t, out, "Libertadores 2018, final")
}

func TestMatchesBySeasonAndDateRange(t *testing.T) {
	s := givenDataLoaded(t)
	out := call(t, s, "search_matches", Args{"team": "Palmeiras", "season": float64(2023)})
	mustContain(t, out, "Palmeiras", "2023")
	from, _, _ := ParseDate("01/06/2019")
	to, _, _ := ParseDate("2019-06-30")
	for _, m := range testDB.Find(MatchFilter{Team: "Santos", From: from, To: to}) {
		if m.Date.Before(from) || m.Date.After(to.Add(24*time.Hour)) {
			t.Errorf("match out of range: %v", m.Date)
		}
	}
	// home venue filter
	for _, m := range testDB.Find(MatchFilter{Team: "Grêmio", Venue: "home", Season: 2018}) {
		if m.HomeKey != "gremio" {
			t.Errorf("home filter returned away match %s", testDB.FormatMatch(m))
		}
	}
}

func TestLastMeeting(t *testing.T) {
	givenDataLoaded(t)
	ms := testDB.Find(MatchFilter{Team: "Flamengo", Opponent: "Corinthians"})
	if len(ms) < 2 || ms[0].Date.Before(ms[1].Date) {
		t.Fatal("matches should be sorted most recent first")
	}
}

func TestTeamNameVariationsGiveSameResults(t *testing.T) {
	givenDataLoaded(t)
	a := len(testDB.Find(MatchFilter{Team: "São Paulo"}))
	for _, v := range []string{"Sao Paulo", "Sao Paulo-SP", "São Paulo - SP", "sao paulo"} {
		if b := len(testDB.Find(MatchFilter{Team: v})); b != a {
			t.Errorf("%q gives %d matches, São Paulo gives %d", v, b, a)
		}
	}
}

func TestStatisticalAnalysis(t *testing.T) {
	s := givenDataLoaded(t)
	out := call(t, s, "competition_stats", Args{"competition": "Brasileirão"})
	mustContain(t, out, "Average goals per match: 2.", "Home win rate:", "Biggest victories:")
	out = call(t, s, "team_rankings", Args{"metric": "away_record", "competition": "Brasileirão"})
	mustContain(t, out, "1. ")
	out = call(t, s, "team_rankings", Args{"metric": "goals_scored", "competition": "Serie A", "season": float64(2019)})
	mustContain(t, out, "1. Flamengo")
	out = call(t, s, "compare_seasons", Args{"seasons": "2018,2019"})
	mustContain(t, out, "2018:", "2019:", "Champion: Palmeiras", "Champion: Flamengo")
}

func TestPlayerQueries(t *testing.T) {
	s := givenDataLoaded(t)
	br := testDB.FindPlayers(PlayerFilter{Nationality: "Brazilian"})
	if len(br) < 500 {
		t.Errorf("only %d Brazilian players", len(br))
	}
	if br[0].Name != "Neymar Jr" {
		t.Errorf("top Brazilian = %s, want Neymar Jr", br[0].Name)
	}
	mustContain(t, call(t, s, "player_details", Args{"name": "Neymar"}), "Neymar Jr", "Paris Saint-Germain", "Dribbling")
	// Gabriel Barbosa is not in FIFA 19; we should get a clear answer with similar names.
	mustContain(t, call(t, s, "player_details", Args{"name": "Gabriel Barbosa"}), "No player named", "Similar names", "Gabriel Jesus")
	mustContain(t, call(t, s, "search_players", Args{"club": "Flamengo"}), "Found 0 players", "not licensed")
	gremio := testDB.FindPlayers(PlayerFilter{Club: "Gremio"})
	if len(gremio) == 0 {
		t.Error("no players at Grêmio (accent-insensitive club search)")
	}
	for _, p := range testDB.FindPlayers(PlayerFilter{Club: "Santos", Position: "forward"}) {
		if !positionMatches("forward", p.Position) {
			t.Errorf("%s is %s", p.Name, p.Position)
		}
	}
	mustContain(t, call(t, s, "brazilian_clubs_players", Args{"brazilian_only": true}), "Brazilian players at Brazilian clubs", "avg rating")
}

// Cross-file query: match record + FIFA players for one club.
func TestCrossFileTeamProfile(t *testing.T) {
	s := givenDataLoaded(t)
	out := call(t, s, "team_profile", Args{"team": "Cruzeiro"})
	mustContain(t, out, "Cruzeiro record", "Competitions for Cruzeiro", "FIFA players at Cruzeiro")
}

func TestDerbiesAndCompetitions(t *testing.T) {
	s := givenDataLoaded(t)
	mustContain(t, call(t, s, "derbies", Args{"season": float64(2023)}), "Fla-Flu", "Grenal")
	mustContain(t, call(t, s, "team_competitions", Args{"team": "Palmeiras"}), "Brasileirão", "Copa do Brasil", "Libertadores")
}

// At least 20 sample questions are answerable without error, fast.
func TestTwentySampleQuestions(t *testing.T) {
	s := givenDataLoaded(t)
	qs := []struct {
		tool string
		args Args
	}{
		{"head_to_head", Args{"team_a": "Flamengo", "team_b": "Fluminense"}},
		{"search_matches", Args{"team": "Palmeiras", "season": 2023.0}},
		{"search_matches", Args{"competition": "Copa do Brasil", "stage": "final"}},
		{"team_stats", Args{"team": "Corinthians", "season": 2022.0, "venue": "home"}},
		{"team_rankings", Args{"metric": "goals_scored", "season": 2023.0, "competition": "Brasileirão"}},
		{"head_to_head", Args{"team_a": "Palmeiras", "team_b": "Santos"}},
		{"search_players", Args{"nationality": "Brazil"}},
		{"search_players", Args{"club": "Cruzeiro"}},
		{"search_players", Args{"club": "Grêmio", "position": "forward"}},
		{"standings", Args{"season": 2019.0}},
		{"search_matches", Args{"competition": "Libertadores", "season": 2018.0, "stage": "semifinals"}},
		{"standings", Args{"season": 2020.0}},
		{"competition_stats", Args{"competition": "Brasileirão"}},
		{"team_rankings", Args{"metric": "away_record"}},
		{"competition_stats", Args{}},
		{"search_matches", Args{"team": "Flamengo", "opponent": "Corinthians", "limit": 1.0}},
		{"player_details", Args{"name": "Gabriel Barbosa"}},
		{"derbies", Args{"season": 2023.0}},
		{"team_competitions", Args{"team": "Palmeiras"}},
		{"team_rankings", Args{"metric": "home_record"}},
		{"search_players", Args{"nationality": "Brazil", "limit": 10.0}},
		{"compare_seasons", Args{"seasons": "2018,2019"}},
		{"team_profile", Args{"team": "Grêmio"}},
		{"search_matches", Args{"team": "Santos", "date_from": "2015-01-01", "date_to": "31/12/2015"}},
	}
	for _, q := range qs {
		if out := call(t, s, q.tool, q.args); len(out) < 20 {
			t.Errorf("%s: suspiciously short output %q", q.tool, out)
		}
	}
}

func TestErrorsReported(t *testing.T) {
	s := givenDataLoaded(t)
	if _, err := s.CallTool("team_stats", Args{"team": "Nonexistent United"}); err == nil {
		t.Error("expected unknown team error")
	}
	if _, err := s.CallTool("nope", Args{}); err == nil {
		t.Error("expected unknown tool error")
	}
}

// MCP protocol round trip over the stdio transport.
func TestMCPProtocol(t *testing.T) {
	s := givenDataLoaded(t)
	in := strings.Join([]string{
		`{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":"2024-11-05","capabilities":{},"clientInfo":{"name":"t","version":"1"}}}`,
		`{"jsonrpc":"2.0","method":"notifications/initialized"}`,
		`{"jsonrpc":"2.0","id":2,"method":"tools/list"}`,
		`{"jsonrpc":"2.0","id":3,"method":"tools/call","params":{"name":"standings","arguments":{"season":2019}}}`,
		`{"jsonrpc":"2.0","id":4,"method":"tools/call","params":{"name":"team_stats","arguments":{"team":"Nowhere FC"}}}`,
		`{"jsonrpc":"2.0","id":5,"method":"bogus"}`,
	}, "\n")
	var out bytes.Buffer
	if err := s.Serve(strings.NewReader(in), &out); err != nil {
		t.Fatal(err)
	}
	lines := strings.Split(strings.TrimSpace(out.String()), "\n")
	if len(lines) != 5 {
		t.Fatalf("got %d responses, want 5 (notification has none):\n%s", len(lines), out.String())
	}
	var resps []map[string]any
	for _, l := range lines {
		var r map[string]any
		if err := json.Unmarshal([]byte(l), &r); err != nil {
			t.Fatal(err)
		}
		resps = append(resps, r)
	}
	if resps[0]["result"].(map[string]any)["protocolVersion"] != protocolVersion {
		t.Error("bad initialize result")
	}
	if n := len(resps[1]["result"].(map[string]any)["tools"].([]any)); n < 10 {
		t.Errorf("only %d tools listed", n)
	}
	res := resps[2]["result"].(map[string]any)
	text := res["content"].([]any)[0].(map[string]any)["text"].(string)
	mustContain(t, text, "Flamengo - 90 pts")
	if resps[3]["result"].(map[string]any)["isError"] != true {
		t.Error("unknown team should be isError")
	}
	if resps[4]["error"] == nil {
		t.Error("unknown method should error")
	}
}

File: experiments/adrianco/experiment-75-opus55-alllang/brazil/runs/effort=low_language=typescript_model=claude-opus-5-5_prompt=neutral/rep1/tests/soccer.test.ts
import { beforeAll, describe, expect, it } from "vitest";
import { loadDataset, type Dataset } from "../src/data.js";
import { createHandlers } from "../src/tools.js";
import { normalizeTeam, parseDate } from "../src/normalize.js";
import { parseCsv } from "../src/csv.js";
import * as Q from "../src/queries.js";

let ds: Dataset;
let h: ReturnType<typeof createHandlers>;
beforeAll(() => { ds = loadDataset(); h = createHandlers(ds); });

describe("Feature: Data loading", () => {
  it("Scenario: all 6 CSV files are loadable", () => {
    expect(Object.keys(ds.fileCounts)).toHaveLength(6);
    expect(ds.fileCounts["Brasileirao_Matches.csv"]).toBeGreaterThan(4000);
    expect(ds.fileCounts["Brazilian_Cup_Matches.csv"]).toBeGreaterThan(1300);
    expect(ds.fileCounts["Libertadores_Matches.csv"]).toBeGreaterThan(1200);
    expect(ds.fileCounts["BR-Football-Dataset.csv"]).toBe(10296);
    expect(ds.fileCounts["novo_campeonato_brasileiro.csv"]).toBe(6886);
    expect(ds.fileCounts["fifa_data.csv"]).toBe(18207);
  });
  it("Scenario: every file is queryable", () => {
    const sources = new Set(Q.findMatches(ds, { team: "Flamengo" }).map((m) => m.source));
    expect(sources.size).toBeGreaterThanOrEqual(3);
    for (const f of ["Brasileirao_Matches.csv", "Brazilian_Cup_Matches.csv", "Libertadores_Matches.csv", "BR-Football-Dataset.csv", "novo_campeonato_brasileiro.csv"])
      expect(ds.matches.some((m) => m.source === f)).toBe(true);
  });
  it("Scenario: UTF-8 accents preserved", () => {
    expect(ds.matches.some((m) => m.home === "Grêmio")).toBe(true);
    expect(ds.players.some((p) => p.club === "Atlético Mineiro")).toBe(true);
  });
});

describe("Feature: Normalization", () => {
  it.each([
    ["Palmeiras-SP", "palmeiras"], ["Palmeiras", "palmeiras"], ["Sport Club Corinthians Paulista", "corinthians"],
    ["São Paulo - SP", "sao paulo"], ["Sao Paulo", "sao paulo"], ["Atlético - MG", "atletico-mg"], ["Atletico Mineiro", "atletico-mg"],
    ["Athletico-PR", "atletico-pr"], ["Atlético Paranaense", "atletico-pr"], ["Vasco Da Gama RJ", "vasco"], ["Gremio RS", "gremio"],
    ["4 de Julho EC", "4 de julho"], ["Botafogo-RJ", "botafogo"], ["Botafogo - SP", "botafogo-sp"], ["Nacional (URU)", "nacional"],
  ])("normalizes %s -> %s", (raw, key) => expect(normalizeTeam(raw)).toBe(key));
  it("handles multiple date formats", () => {
    expect(parseDate("2023-09-24")).toBe("2023-09-24");
    expect(parseDate("29/03/2003")).toBe("2003-03-29");
    expect(parseDate("2012-05-19 18:30:00")).toBe("2012-05-19");
    expect(parseDate("Jul 1, 2004")).toBe("2004-07-01");
  });
  it("parses quoted CSV", () => {
    expect(parseCsv('a,b\n"x, y","say ""hi"""\n')).toEqual([{ a: "x, y", b: 'say "hi"' }]);
  });
});

describe("Feature: Match Queries", () => {
  it("Scenario: Find matches between two teams", () => {
    // Given the match data is loaded / When I search Flamengo vs Fluminense
    const ms = Q.findMatches(ds, { team: "Flamengo", opponent: "Fluminense" });
    // Then I receive a list of matches with date, scores and competition
    expect(ms.length).toBeGreaterThan(20);
    for (const m of ms) {
      expect(m.date).toMatch(/^\d{4}-\d{2}-\d{2}$/);
      expect(Number.isInteger(m.homeGoals) && Number.isInteger(m.awayGoals)).toBe(true);
      expect(m.competition).toBeTruthy();
      expect([m.homeKey, m.awayKey].sort()).toEqual(["flamengo", "fluminense"]);
    }
    expect(h.head_to_head({ teamA: "Flamengo", teamB: "Fluminense" })).toMatch(/Fla-Flu derby[\s\S]*Head-to-head in dataset/);
  });
  it("Scenario: matches by team and season", () => {
    const ms = Q.findMatches(ds, { team: "Palmeiras", season: 2023 });
    expect(ms.length).toBeGreaterThan(38);
    expect(ms.every((m) => m.season === 2023)).toBe(true);
  });
  it("Scenario: no duplicates across overlapping files", () => {
    // 2022 is present in both Brasileirao_Matches.csv and BR-Football-Dataset.csv
    const ms = Q.findMatches(ds, { team: "Palmeiras", season: 2022, competition: "Brasileirão" });
    expect(ms.length).toBe(38);
  });
  it("Scenario: Copa do Brasil finals", () => {
    const ms = Q.findMatches(ds, { competition: "Copa do Brasil", stage: "final" });
    expect(ms.length).toBeGreaterThan(0);
    expect(ms.some((m) => m.season === 2020 && [m.homeKey, m.awayKey].includes("palmeiras"))).toBe(true);
  });
  it("Scenario: date range and venue filters", () => {
    const ms = Q.findMatches(ds, { team: "Santos", venue: "home", dateFrom: "2015-01-01", dateTo: "2015-12-31" });
    expect(ms.length).toBeGreaterThan(0);
    expect(ms.every((m) => m.homeKey === "santos" && m.date.startsWith("2015"))).toBe(true);
  });
  it("Scenario: last Flamengo vs Corinthians match", () => {
    const [last] = Q.findMatches(ds, { team: "Flamengo", opponent: "Corinthians" });
    expect(last.date >= "2023-01-01").toBe(true);
  });
  it("Scenario: derbies in a season", () => {
    const d = Q.derbies(ds, { season: 2019 });
    expect(d.map((x) => x.derby)).toContain("Grenal");
  });
});

describe("Feature: Team Queries", () => {
  it("Scenario: Get team statistics", () => {
    // When I request statistics for Palmeiras in season 2023
    const { record } = Q.teamRecord(ds, "Palmeiras", { season: 2023, competition: "Brasileirão" });
    expect(record.matches).toBeGreaterThanOrEqual(37); // 2023 comes from BR-Football-Dataset only, which misses a fixture
    expect(record.wins + record.draws + record.losses).toBe(record.matches);
    expect(record.goalsFor).toBeGreaterThan(0);
    expect(h.team_stats({ team: "Palmeiras", season: 2023 })).toMatch(/Wins: \d+, Draws: \d+, Losses: \d+/);
  });
  it("Scenario: home record", () => {
    const out = h.team_stats({ team: "Corinthians", season: 2022, venue: "home", competition: "Brasileirão" });
    expect(out).toContain("Corinthians home record");
    expect(out).toMatch(/Win rate: \d+\.\d%/);
  });
  it("Scenario: competitions a team played in", () => {
    const comps = Q.teamCompetitions(ds, "Palmeiras").map((c) => c.competition);
    expect(comps).toEqual(expect.arrayContaining(["Brasileirão Série A", "Copa do Brasil", "Copa Libertadores"]));
  });
  it("Scenario: ambiguous team name is flagged", () => {
    expect(h.team_stats({ team: "Atletico" })).toContain("matches several clubs");
  });
});

describe("Feature: Competition Queries", () => {
  it("Scenario: Who won the 2019 Brasileirão?", () => {
    const t = Q.standings(ds, 2019);
    expect(t).toHaveLength(20);
    expect(t[0].team).toBe("Flamengo");
    expect(t[0].points).toBe(90);
    expect(h.standings({ season: 2019 })).toMatch(/1\. Flamengo - 90 pts \(28W, 6D, 4L.*Champion/);
  });
  it("Scenario: relegated teams in 2020", () => {
    const out = h.standings({ season: 2020 });
    const relegated = out.split("\n").filter((l) => l.includes("Relegated"));
    expect(relegated).toHaveLength(4);
    expect(out).toMatch(/Botafogo.*Relegated/);
  });
  it("Scenario: historical season from 2003-2019 file", () => {
    expect(Q.standings(ds, 2005).length).toBeGreaterThanOrEqual(20);
  });
  it("Scenario: Libertadores stage", () => {
    const out = h.search_matches({ competition: "Libertadores", season: 2018, stage: "final" });
    expect(out).toMatch(/final/);
  });
});

describe("Feature: Statistical Analysis", () => {
  it("Scenario: average goals per match", () => {
    const s = Q.competitionStats(ds, { competition: "Brasileirão" });
    expect(s.avgGoals).toBeGreaterThan(2);
    expect(s.avgGoals).toBeLessThan(3.5);
    expect(h.competition_stats({ competition: "Brasileirão" })).toMatch(/Average goals per match: \d\.\d\d/);
  });
  it("Scenario: biggest wins sorted by margin", () => {
    const ws = Q.biggestWins(ds, { limit: 10 });
    const margins = ws.map((m) => Math.abs(m.homeGoals - m.awayGoals));
    expect([...margins].sort((a, b) => b - a)).toEqual(margins);
    expect(margins[0]).toBeGreaterThanOrEqual(7);
  });
  it("Scenario: best away record", () => {
    expect(h.best_records({ venue: "away", competition: "Brasileirão" })).toMatch(/^Best away records[\s\S]*1\. /);
  });
  it("Scenario: team scored the most goals in a season", () => {
    const s = Q.competitionStats(ds, { competition: "Brasileirão", season: 2019 });
    expect(s.topScoringTeams[0].team).toBe("Flamengo");
  });
});

describe("Feature: Player Queries", () => {
  it("Scenario: top Brazilian players", () => {
    const ps = Q.searchPlayers(ds, { nationality: "Brazilian", limit: 3 });
    expect(ps[0].name).toBe("Neymar Jr");
    expect(ps.every((p) => p.nationality === "Brazil")).toBe(true);
  });
  it("Scenario: players at a club (accent-insensitive)", () => {
    const ps = Q.searchPlayers(ds, { club: "Gremio", limit: 100 });
    expect(ps.length).toBe(20);
    expect(ps.every((p) => p.club === "Grêmio")).toBe(true);
  });
  it("Scenario: club match is not a substring match", () => {
    expect(Q.searchPlayers(ds, { club: "Santos", limit: 100 }).every((p) => p.club === "Santos")).toBe(true);
  });
  it("Scenario: forwards by position group", () => {
    const ps = Q.searchPlayers(ds, { club: "Cruzeiro", position: "forward", limit: 100 });
    expect(ps.length).toBeGreaterThan(0);
    expect(ps.every((p) => ["ST", "CF", "LF", "RF", "LS", "RS", "LW", "RW"].includes(p.position))).toBe(true);
  });
  it("Scenario: player by name", () => {
    expect(h.get_player({ name: "Neymar" })).toMatch(/Neymar Jr[\s\S]*Overall: 92/);
    expect(h.get_player({ name: "Nobody Atall" })).toMatch(/No player/);
  });
  it("Scenario: cross-file query (players + matches)", () => {
    const rows = Q.brazilianClubPlayers(ds, { brazilianOnly: true });
    expect(rows.map((r) => r.club)).toContain("Cruzeiro");
    const profile = h.team_profile({ team: "Cruzeiro" });
    expect(profile).toMatch(/all-time record[\s\S]*Brasileirão Série A[\s\S]*Overall: \d+/);
  });
});

describe("Feature: Performance and robustness", () => {
  it("answers 20+ sample questions each within 2s", () => {
    const calls: [string, any][] = [
      ["search_matches", { team: "Flamengo", opponent: "Fluminense" }], ["search_matches", { team: "Palmeiras", season: 2023 }],
      ["search_matches", { competition: "Copa do Brasil", stage: "final" }], ["team_stats", { team: "Corinthians", season: 2022, venue: "home" }],
      ["competition_stats", { competition: "Série A", season: 2023 }], ["head_to_head", { teamA: "Palmeiras", teamB: "Santos" }],
      ["search_players", { nationality: "Brazil" }], ["search_players", { club: "Fluminense" }], ["search_players", { club: "São Paulo", position: "forward" }],
      ["standings", { season: 2019 }], ["search_matches", { competition: "Libertadores", season: 2018 }], ["standings", { season: 2020 }],
      ["competition_stats", { competition: "Brasileirão" }], ["best_records", { venue: "away" }], ["biggest_wins", {}],
      ["search_matches", { team: "Flamengo", opponent: "Corinthians", limit: 1 }], ["get_player", { name: "Gabriel" }],
      ["derbies", { season: 2023 }], ["team_competitions", { team: "Palmeiras" }], ["best_records", { venue: "home" }],
      ["standings", { season: 2018 }], ["team_profile", { team: "Internacional" }], ["brazilian_club_players", {}], ["dataset_info", {}],
    ];
    for (const [n, a] of calls) {
      const t = performance.now();
      const out = h[n](a);
      expect(performance.now() - t).toBeLessThan(2000);
      expect(out.length).toBeGreaterThan(10);
    }
  });
  it("rejects unknown competitions clearly", () => {
    expect(() => h.search_matches({ competition: "Premier League" })).toThrow(/Unknown competition/);
  });
});
